### ISSUE TITLE
[feat req] Use in browser via webpack, browserify

### ISSUE THREAD

**Issue by redbrain** [PrismarineJS/flying-squid#491]
Hello,
It would be nice if flying-squid would be webpackable to use completely client-side.
What currently limits flying-squid from being webpacked, and can these limitations be ignored to create a simple server?

**Comment by rom1504** [comment]
yeah totally agree, that's something we want for https://github.com/PrismarineJS/prismarine-web-client/issues/47

Nobody tried to do it but the minimum we need is a connection provider in node-minecraft-protocol server to replace the tcp server. Opened an issue for this there https://github.com/PrismarineJS/node-minecraft-protocol/issues/829

**Comment by redbrain** [comment]
Good to hear that this is already in consideration.
I'll subscribe to the mentioned issues, and speak here again once those are mature enough to implement a client-side server.

**Comment by redbrain** [comment]
In addition to a TCP alternative, we'll need to provide an alternative to FS, since browsers don't have native FS access (yet).
For player data, perhaps the browser's localStorage is a good alternative; that will probably need an option to be enabled.
The items exported via requireindex could probably be specified manually, at least to my understanding.
I'm not sure how we should implement world storage and logging. I also haven't looked at FS usage in any dependencies.

**Comment by rom1504** [comment]
Yeah world storage can be done using indexeddb

**Comment by mollthecoder** [comment]
BrowserFS would be a good way to access FS in browser.

**Comment by hard-shutdown** [comment]
Maybe WebSockets for the client communication?

**Comment by extremeheat** [comment]
No network connection is necessary in single player mode

**Comment by hard-shutdown** [comment]
> No network connection is necessary in single player mode

Fair, we could use a one-sided messaging channel with some javascript, like the way socket.io works for messages.

**Comment by extremeheat** [comment]
All that needs to be done is use the existing event emitters. Drop nmp dep and when mineflayer does a write, the event should instead be emitted to flying squid directly.

**Comment by extremeheat** [comment]
No serialization is needed, and when flying squid wants to write it would just go right to mineflayer bot eventemitter

**Comment by tutacat** [comment]
webStorage isn't an _actual_ fs, it is emulated, and indexedDB is much faster, and with a Blob

But does it even make sense to webpack it???

**Comment by vantezzen** [comment]
I've worked on wrapping flying-squid to let it run in the browser over at https://github.com/vantezzen/electric-squid (https://squid.blymp.io/). It's still a very rough proof-of-concept, but connecting to the server via a Minecraft Client is still (mostly) possible.

There were some difficulties getting webpack to build the bundle (mainly solved by increasing the RAM limit, writing custom polyfills and patching dependencies manually) but generally it seems do-able. webpack's tree-shaking also seems to be very stressed as some small changes to the code result in the bundle size fluctuating between 50-400MB.

I've used the methods discussed here to create a client-server connection via WebSockets and using IndexedDB via the level filesystem to replace the fs module.

**Comment by rom1504** [comment]
@vantezzen that's awesome!
Let's us know how it goes
If you see changes that could make this experiment easier, do say

**Comment by zardoy** [comment]
Hi everyone! I couldn't find a way to polyfill all the necessary components, so I decided to debug and modify the required code instead.
Right now, I've managed to use this server in the browser and implemented the desired singleplayer support in [forked mineflayer web client](https://github.com/zardoy/prismarine-web-client). In the browser, we can easily save worlds in browser memory or interact with the real file system instead (however the sad part is that Chrome doesn't allow to open folders in %appdata% directly). I don't have a clean solution yet, though there are already some improvements in the [fork](github.com/zardoy/space-squid)
